clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/AddExecutor.scala

```diff
@@ -42,8 +46,142 @@ class AddExecutor[TI <: TransferIndex](addCommand: AddCommand[TI])
       Future
         .failed(_) logErrorMsg s"Metadata at $location cannot be added to Clio", {
         decoded =>
-          webClient.upsert(addCommand.index)(addCommand.key, decoded)
+          {
+            for {
+              existingMetadata <- if (!addCommand.forceUpdate) {
+                queryForKey(webClient).logErrorMsg(
+                  s"Could not query the $name. No files have been added."
+                )
+              } else {
+                Future.successful(None)
+              }
+              upsertResponse <- addFiles(
+                webClient,
+                decoded,
+                existingMetadata
+              ).fold(
+                left => Future.failed(left).logErrorMsg(left.getMessage),
+                identity
+              )
+            } yield {
+              upsertResponse
+            }
+          }
       }
     )
   }
+
+  private def addFiles(
+    client: ClioWebClient,
+    newMetadata: addCommand.index.MetadataType,
+    existingMetadata: Option[addCommand.index.MetadataType],
+  )(implicit ec: ExecutionContext): Either[Throwable, Future[UpsertId]] = {
+    val differences = existingMetadata.fold(Iterable.empty[(String, Any, Any)]) {
+      diffMetadata(_, newMetadata)
+    }
+    for {
+      upsertResponse <- Either.cond(
+        differences.isEmpty,
+        client
+          .upsert(addCommand.index)(addCommand.key, newMetadata)
+          .logErrorMsg(
+            s"An error occurred while adding the $name record in Clio."
+          ),
+        new RuntimeException(
+          "Adding this document will overwrite the following existing metadata: " +
+            differences
+              .map(
+                diff => s"Field: ${diff._1} Old value: ${diff._3} New value: ${diff._2}"
+              )
+              .mkString("\n") +
+            " Use '--force-update' to overwrite this data."
+        )
+      )
+
+    } yield {
+      upsertResponse
+    }
+  }
+
+  private def diffMetadata(
+    existingMetadata: addCommand.index.MetadataType,
+    newMetadata: addCommand.index.MetadataType
+  ): Iterable[(String, Any, Any)] = {
+    val mapper = new CaseClassMapper[addCommand.index.MetadataType]
+
+    def filterNones(value: (String, Any)) = {
+      value._2 match {
```

Comment:
Does `!value._2.equals(None)` work? If so we can probably just inline it into the `filter` calls.